keps/sig-node/20190226-migrate-in-tree-dockershim-to-out-of-tree.md

```diff
@@ -0,0 +1,158 @@
+
+### Pros
+- Docker is not special and just a CRI just like every other CRI in our ecosystem.
+- Currently, dockershim "enjoys" some backdoors for various reasons. Deprecating these "features" should eliminate maintenance burden of kubelet.
+- A cri-dockerd can be maintained independently.
+- Over time we can remove vendored docker dependencies in kubelet.
+
+Having said that, cons of deprecation built-in dockershim requires lots of attentions:
+
+### Cons
+- Deployment pain with a new binary in addition to kubelet.
+  - An additional component may aggravate the complexity currently. It may be relieved with docker version evolutions.
+- The number of affected users maybe large.
+  - Users must change existing use experience when using Kubernetes and docker.
+  - Users have to change their existing workflows to adapt to this new changes.
+  - And other unrecorded stuff.
+- Updating all the eco-system tools to support the new cri-dockerd.
+- Many people use the built in dockershim for in-cluster image build. While that may not be something we recommend for a variety of reasons, it will be a breaking change for these users.
+- CRI is still in alpha，should probably get a 1.0 out there splitting out dockershim completely from kubelet.
+- Existing CNI and CSI plugins may also be affected.
+  - Current dockershim has independent module interacting with CNI plugins. After migrating dockershim out of Kubelet, it may affect some processes between dockershim and CNI plugins.
+- cri-dockerd will vendor kubernetes/kubernetes, that may be tough.
+- cri-dockerd as an independent software running on node should be allocated enough resource to guarantee its availability.
+
+> You can check [the discussion in sig-node mailing list](https://groups.google.com/forum/#!msg/kubernetes-sig-node/0qVzfugYhro/l6Au216XAgAJ) for more details. 
+
+Based on all the discussion, we agree that we should not rush to immediate decision. At the same time, it's the right time to start designing and documenting dockershim deprecation criteria and plan, which will be the main content of rest of this KEP.
+
+### Goals
+
+- A concrete dockershim deprecation criteria.
+- A brief plan to deprecate dockershim spanning multiple releases.
+
+### Non-Goals
+
+- Deprecation of dockershim immediately without consideration for users and WIP efforts depending on it.
+- Refactoring or re-design of dockershim itself due to deprecation.
+
+## Proposal
+
+### Dockershim deprecation criteria
+
+- CRI itself is beta.
+- kubelet has no dependency on dockershim/docker in its whole lifecycle. 
+- All node related features are CRI generic and have no "back door" dependency on dockershim/docker.
+- Deprecate and remove, or replace all Docker-specific features.
+- Reasonable benchmark result of performance degradation after moving dockershim to out-of-tree.
```

Comment:
What percentage of degradation is acceptable (10%) ?

Reply:
10% might not be acceptable by the wider audience.
but my guess is that it's going to end up being less than that 🤔